compiler_opt/tools/combine_tfa_policies_lib.py

```
@@ -0,0 +1,138 @@
+    super().__init__(*args, **kwargs)
+
+    self.tf_policies = []
+    self.tf_policy_names = []
+    for name, policy in tf_policies.items():
+      self.tf_policies.append(policy)
+      self.tf_policy_names.append(name)
+
+    self.expected_signature = self.time_step_spec
+    self.sorted_keys = sorted(self.expected_signature.observation.keys())
+
+    high_low_tensors = []
+    for name in self.tf_policy_names:
+      m = hashlib.md5()
+      m.update(name.encode("utf-8"))
+      high_low_tensors.append(
+          tf.stack([
+              tf.constant(
+                  int.from_bytes(m.digest()[8:], "little"), dtype=tf.uint64),
+              tf.constant(
+                  int.from_bytes(m.digest()[:8], "little"), dtype=tf.uint64)
+          ]))
+    self.high_low_tensors = tf.stack(high_low_tensors)
+    # Related LLVM commit: https://github.com/llvm/llvm-project/pull/96276
+    m = hashlib.md5()
+    m.update(self.tf_policy_names[0].encode("utf-8"))
+    self.high = int.from_bytes(m.digest()[8:], "little")
+    self.low = int.from_bytes(m.digest()[:8], "little")
+    self.high_low_tensor = tf.constant([self.high, self.low], dtype=tf.uint64)
+
+  def _process_observation(
+      self, observation: types.NestedSpecTensorOrArray
+  ) -> Tuple[types.NestedSpecTensorOrArray, types.TensorOrArray]:
+    assert "model_selector" in self.sorted_keys
+    high_low_tensor = self.high_low_tensor
+    for name in self.sorted_keys:
+      if name in ["model_selector"]:
+        switch_tensor = observation.pop(name)[0]
+        high_low_tensor = switch_tensor
+
+        tf.debugging.Assert(
+            tf.equal(
+                tf.reduce_any(
+                    tf.reduce_all(
+                        tf.equal(high_low_tensor, self.high_low_tensors),
+                        axis=1)), True),
+            [high_low_tensor, self.high_low_tensors])
+
+    return observation, high_low_tensor
+
+  def _create_distribution(self, inlining_prediction):
```

open q: can we document this isn't introducing a new behavior? or is there some way to delegate the implementation? (OK to document and add a TODO if no easy way to delegate. If you go the TODO path, please open an issue and tag it here, i.e. TODO(1234)..)

---

This could introduce new behavior, e.g., when combining a deterministic policy such as the saved ppo models. Such policies do not return a proper distribution, however, this is not a good behavior IMO as we use distributions in our trajectory collecting code. In general (my) preferred behavior is to always return a discrete distribution over the action space even if that distribution is supported only on a single action, i.e., deterministic distributions will always look like [1, 0] or [0, 1]. Note that this will only create a new behavior when calling distribution. The behavior of action is remains unchanged.

---

> Note that this will only create a new behavior when calling distribution. The behavior of action is remains unchanged.

This means `_create_distribution` isn't used for inference? Can we raise an exception then or something that says "not implemented, and intentionally so"?

---

Raising exceptions in tf did not work in a saved model as you experienced the first time you used the combined policy and didn't pass the right model_selector name. I can remove this function and let _distribution just return something trivial or leave it empty (not sure if this would work), however, if I remember correctly every tf policy needs some type of distribution function.

---

Ack on there needing to be a function.

Is it correct that the function isn't actually used during inference, and then it's just that we can't just throw (because of how tf evaluates it)... can we make it return something _invalid_ (and document what's going on, too)? Basically, we want to (1) be clear this is just an artifact of the framework (tf_agents), not actually needed, (2) make it fail-fast if our assumptions in (1) are invalidated in the future, which helps both maintaining the code, and a reader of the code.

---

Set a default deterministic distribution as discussed.